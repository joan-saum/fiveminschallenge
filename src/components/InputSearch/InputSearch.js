// @flow

import React, {useCallback, useEffect, useState} from 'react';
import {View, Text, TextInput} from 'react-native';
import type {Node} from 'react';

import styles from './InputSearch.style';

type Props = {|
  onSearch: (search: string) => void,
|};

let timeoutIDSearchMovie: TimeoutID | null = null;
const TIME_BEFORE_FETCH = 2000;

const InputSearch = ({onSearch}: Props): Node => {
  const [text, onChangeText] = useState('');

  const onFilterChanged = useCallback(
    (search: string) => {
      clearTimeout(timeoutIDSearchMovie);

      timeoutIDSearchMovie = setTimeout(() => {
        onSearch(search);
      }, TIME_BEFORE_FETCH);

      onChangeText(search);
    },
    [onChangeText, onSearch],
  );

  useEffect(() => {
    return () => clearTimeout(timeoutIDSearchMovie);
  }, []);

  return (
    <View style={styles.container}>
      <Text>Search:</Text>
      <TextInput
        onChangeText={onFilterChanged}
        value={text}
        style={styles.input}
      />
    </View>
  );
};

export default InputSearch;
