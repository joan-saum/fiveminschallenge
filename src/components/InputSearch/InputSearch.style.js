import {StyleSheet} from 'react-native';

const style = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    marginLeft: 20,
    marginRight: 20,
    marginBottom: 20,
    paddingLeft: 10,
    paddingRight: 10,
    borderWidth: 1,
  },
  input: {
    alignItems: 'center',
    // borderWidth: 1,
    flexGrow: 1,
    // height: 35,
    color: 'black',
  },
});

export default style;
