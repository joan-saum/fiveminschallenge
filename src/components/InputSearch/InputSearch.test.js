// @flow

import * as React from 'react';
import renderer, {act} from 'react-test-renderer';
import {TextInput} from 'react-native';

import InputSearch from './InputSearch';

jest.unmock('./InputSearch');

jest.useFakeTimers();

describe('InputSearch', () => {
  it('should render a InputSearch', () => {
    const onSearch = jest.fn().mockName('onSearch');
    const tree = renderer.create(<InputSearch onSearch={onSearch} />);

    expect(tree).toMatchSnapshot();
  });

  it('should call onSearch only after 1 second without changes', () => {
    const onSearch = jest.fn().mockName('onSearch');
    const tree = renderer.create(<InputSearch onSearch={onSearch} />);

    expect(onSearch).toHaveBeenCalledTimes(0);

    act(() => {
      tree.root.findByType(TextInput).props.onChangeText('test');

      expect(onSearch).toHaveBeenCalledTimes(0);
      jest.advanceTimersByTime(500);
      expect(onSearch).toHaveBeenCalledTimes(0);

      tree.root.findByType(TextInput).props.onChangeText('test2');
      expect(onSearch).toHaveBeenCalledTimes(0);
      jest.advanceTimersByTime(500);
      expect(onSearch).toHaveBeenCalledTimes(0);
      jest.advanceTimersByTime(500);
      expect(onSearch).toHaveBeenCalledTimes(1);
      expect(onSearch).toHaveBeenCalledWith('test2');
    });

    tree.unmount();
  });
});
