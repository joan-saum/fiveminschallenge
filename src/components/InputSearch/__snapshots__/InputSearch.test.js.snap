// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSearch should render a InputSearch 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "borderWidth": 1,
      "display": "flex",
      "flexDirection": "row",
      "marginBottom": 20,
      "marginLeft": 20,
      "marginRight": 20,
      "paddingLeft": 10,
      "paddingRight": 10,
    }
  }
>
  <Text>
    Search:
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Object {
        "alignItems": "center",
        "color": "black",
        "flexGrow": 1,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
